File: Aulas/Aula06/PadraoFacade.java
class Cafeteira {
    public void moerGraos() {
        System.out.println("Moendo graos de cafe");
    }

    public void fazerCafe() {
        System.out.println("Fazendo cafe");
    }
}

class Chaleira {
    public void ferverAgua() {
        System.out.println("Fervendo agua");
    }

    public void fazerCha() {
        System.out.println("Fazendo cha");
    }
}

class BebidasF {
    private Cafeteira cafeteira;
    private Chaleira chaleira;

    public BebidasF() {
        this.cafeteira = new Cafeteira();
        this.chaleira = new Chaleira();
    }

    public void prepararCafe() {
        System.out.println("\nPreparando cafe...");
        cafeteira.moerGraos();
        cafeteira.fazerCafe();
    }

    public void prepararCha() {
        System.out.println("\nPreparando cha...");
        chaleira.ferverAgua();
        chaleira.fazerCha();
    }
}

public class PadraoFacade {
    public static void main(String[] args) {
        BebidasF facade = new BebidasF();
        facade.prepararCafe();
        facade.prepararCha();
    }
}

File: Aulas/Aula06/Padrao_facade.go
package main

import "fmt"

type Cafeteira struct{}

func (c Cafeteira) MoerGraos() {
	fmt.Println("Moendo grãos de cafe")
}

func (c Cafeteira) FazerCafe() {
	fmt.Println("Fazendo cafe")
}

type Chaleira struct{}

func (c Chaleira) FerverAgua() {
	fmt.Println("Fervendo agua")
}

func (c Chaleira) FazerCha() {
	fmt.Println("Fazendo cha")
}

type BebidasFacade struct {
	cafeteira Cafeteira
	chaleira  Chaleira
}

func NewBebidasFacade() BebidasFacade {
	return BebidasFacade{
		cafeteira: Cafeteira{},
		chaleira:  Chaleira{},
	}
}

func (b BebidasFacade) PrepararCafe() {
	fmt.Println("\nPreparando cafe...")
	b.cafeteira.MoerGraos()
	b.cafeteira.FazerCafe()
}

func (b BebidasFacade) PrepararCha() {
	fmt.Println("\nPreparando cha...")
	b.chaleira.FerverAgua()
	b.chaleira.FazerCha()
}

func main() {
	facade := NewBebidasFacade()
	facade.PrepararCafe()
	facade.PrepararCha()
}
